Bug 487750 - test_plugin_clipping.xhtml fails randomly. r+sr=bz

diff --git a/layout/generic/test/plugin_clipping_helper.xhtml b/layout/generic/test/plugin_clipping_helper.xhtml
--- a/layout/generic/test/plugin_clipping_helper.xhtml
+++ b/layout/generic/test/plugin_clipping_helper.xhtml
@@ -159,19 +159,31 @@ function checkClipRegion(id, rects) {
 
 function checkClipRegionForFrame(fid, id, rects) {
   var f = document.getElementById(fid);
   var bounds = f.getBoundingClientRect();
   checkClipRegionWithDoc(f.contentDocument, bounds.left, bounds.top, id, rects);
 }
 
 function runTests2() {
-  var p = document.getElementById("p1");
-  if (p.getClipRegionRectEdge(0, 0) == p.getClipRegionRectEdge(0, 2)) {
-    // plugin hasn't been updated yet. wait.
+  function check(id, doc) {
+    var p = (doc || document).getElementById(id);
+    if (p.getClipRegionRectEdge(0, 0) ==
+        p.getClipRegionRectEdge(0, 2))
+      throw id;
+  }
+
+  try {
+    check("p1");
+    check("p2");
+    check("p3");
+    check("p4");
+    check("p5", f1.contentDocument);
+  } catch (id) {
+    // One or more plugins haven't been updated yet. Wait.
     setTimeout(runTests2, 100);
     return;
   }
 
   checkClipRegion("p1", [[0, 0, 200, 200]]);
   checkClipRegion("p2", [[100, 100, 200, 200]]);
   checkClipRegion("p3", [[100, 100, 200, 200]]);
   checkClipRegion("p4", [[100, 100, 200, 200]]);
