diff --git a/xpcom/glue/FunctorQueue.h b/xpcom/glue/FunctorQueue.h
new file mode 100644
--- /dev/null
+++ b/xpcom/glue/FunctorQueue.h
@@ -0,0 +1,213 @@
+#ifndef __functorqueue_h__
+#define __functorqueue_h__
+
+#include "nsISupports.h"
+#include "nsAutoPtr.h"
+#include "nsAutoLock.h"
+#include "nsIEventTarget.h"
+#include "nsThreadUtils.h"
+#include "Functor.h"
+
+#include <vector>
+
+
+class PayloadBase
+{
+  friend class FunctorQueue;
+
+  nsAutoRefCnt mRefCnt;
+  PRMonitor* const mMon;
+  PRBool mDelivered;
+
+public:
+
+  PayloadBase()
+    : mMon(nsAutoMonitor::NewMonitor("PayloadBase"))
+    , mDelivered(PR_FALSE)
+  {}
+
+  NS_IMPL_THREADSAFE_ADDREF(PayloadBase)
+  NS_IMPL_THREADSAFE_RELEASE(PayloadBase)
+  void VoidRelease() { Release(); }
+
+  void Wait()
+  {
+    nsAutoMonitor mon(mMon);
+    if (!mDelivered)
+      mon.Wait();
+  }
+
+  virtual ~PayloadBase()
+  {
+    nsAutoMonitor::DestroyMonitor(mMon);
+  }
+
+protected:
+
+  virtual void Deliver() = 0;
+  void PostDelivery()
+  {
+    nsAutoMonitor mon(mMon);
+    mDelivered = PR_TRUE;
+    mon.NotifyAll();
+  }
+
+};
+
+
+template <typename Result>
+class Payload : public PayloadBase
+{
+  Result mResult;
+  typedef Functor<Result, TLIST0()> functor_type;
+  functor_type mFun;
+
+public:
+
+  Payload(functor_type fun) : mFun(fun) {}
+
+  Result& Wait()
+  {
+    PayloadBase::Wait();
+    return mResult;
+  }
+
+protected:
+
+  void Deliver() {
+    mResult = mFun();
+    PostDelivery();
+  }
+
+};
+
+
+template <>
+class Payload<void> : public PayloadBase
+{
+  typedef Functor<void, TLIST0()> functor_type;
+  functor_type mFun;
+
+public:
+
+  Payload(functor_type fun) : mFun(fun) {}
+
+protected:
+
+  void Deliver() {
+    mFun();
+    PostDelivery();
+  }
+
+};
+
+
+class FunctorQueue
+{
+  typedef std::vector<PayloadBase*> queue_type;
+
+  PRLock* mLock;
+  queue_type mQueue;
+
+  nsAutoRefCnt mRefCnt;
+
+public:
+
+  FunctorQueue()
+    : mLock(nsAutoLock::NewLock("FunctorQueue"))
+  {}
+
+  NS_IMPL_THREADSAFE_ADDREF(FunctorQueue)
+  NS_IMPL_THREADSAFE_RELEASE(FunctorQueue)
+
+  virtual ~FunctorQueue()
+  {
+    NS_ASSERTION(mQueue.size() == 0, "Deleting non-empty FunctorQueue");
+    nsAutoLock::DestroyLock(mLock);
+  }
+
+  virtual void Flush()
+  {
+    queue_type temp;
+
+    { nsAutoLock sync(mLock);
+      temp = mQueue;
+      mQueue.clear(); }
+
+    PRBool isMainThread = NS_IsMainThread();
+
+    queue_type::const_iterator end = temp.end();
+    for (queue_type::iterator it = temp.begin(); end != it; ++it)
+    {
+      PayloadBase* const pl = *it;
+      pl->Deliver();
+
+      if (isMainThread)
+        pl->Release();
+      else // Proxy Release to the main thread:
+        NS_DispatchToMainThread(NS_NEW_RUNNABLE_METHOD(
+          PayloadBase, pl, VoidRelease));
+    }
+  }
+
+  // Despite the name, subclasses may implement AsyncFlush synchronously:
+  virtual void AsyncFlush() = 0;
+
+  template <typename Result>
+  nsRefPtr<Payload<Result> >
+  Enqueue(Functor<Result, TLIST0()> f)
+  {
+    nsRefPtr<Payload<Result> >
+      pl(new Payload<Result>(f));
+    NS_ENSURE_TRUE(pl, 0);
+
+    // This AddRef balances the Release call in Flush
+    // TODO could use the unsafe AddRef implementation
+    pl->AddRef();
+
+    { nsAutoLock sync(mLock);
+      mQueue.push_back(pl); }
+
+    AsyncFlush();
+
+    // pl's refcnt still >= 1, even if already Flushed
+    return pl;
+  }
+
+};
+
+class nsFunctorQueue
+  : public FunctorQueue
+{
+  nsRefPtr<nsIEventTarget> mTarget;
+  nsIRunnable* mPendingEvent;
+
+public:
+
+  nsFunctorQueue(nsIEventTarget* target)
+    : mTarget(target)
+    , mPendingEvent(nsnull)
+  {}
+
+  void Flush() {
+    mPendingEvent = nsnull;
+    FunctorQueue::Flush();
+  }
+
+  void AsyncFlush()
+  {
+    PRBool sameThread;
+    // TODO IsOnCurrentThread not safe for threadpools!!!
+    if (NS_SUCCEEDED(mTarget->IsOnCurrentThread(&sameThread)) && !sameThread) {
+      if (mPendingEvent)
+        return;
+      mPendingEvent = NS_NEW_RUNNABLE_METHOD(nsFunctorQueue, this, Flush);
+      mTarget->Dispatch(mPendingEvent, NS_DISPATCH_NORMAL);
+    } else {
+      Flush();
+    }
+  }
+
+};
+
+#endif
\ No newline at end of file
diff --git a/xpcom/glue/Makefile.in b/xpcom/glue/Makefile.in
--- a/xpcom/glue/Makefile.in
+++ b/xpcom/glue/Makefile.in
@@ -129,6 +129,7 @@
 		TList.h \
 		Functor.h \
 		Bind.h \
+		FunctorQueue.h \
 		$(NULL)
 
 SDK_LIBRARY     =                        \
