diff --git a/xpcom/glue/Functor.h b/xpcom/glue/Functor.h
--- a/xpcom/glue/Functor.h
+++ b/xpcom/glue/Functor.h
@@ -38,16 +38,17 @@
 
 #ifndef Functor_h__
 #define Functor_h__
 
 #include "nsISupports.h"
 #include "nsIRunnable.h"
 #include "nsAutoPtr.h"
 #include "TList.h"
+#include "nsRunnableFunctor.h"
 
 /**
  * For an overview of Functor usage, see bug 486440.
  *
  * This implementation was inspired by Andrei Alexandrescu's book
  *
  *   Modern C++ Design: Generic Programming and Design Patterns Applied,
  *
@@ -209,16 +210,20 @@ public:
 
   EXTRACT_TYPEDEFS(Params)
 
 public:
   Functor(const Functor& f) : mImpl(f.mImpl) {}
   Functor(ImplType* impl)   : mImpl(impl) {}
  ~Functor() {}
 
+  nsRefPtr<nsRunnableFunctor<Result, Functor> >
+  proxy(nsIThread* target,
+        PRUint32 flags = NS_DISPATCH_SYNC);
+
   template <typename Callable>
   static Functor callable(Callable c) {
     return Functor(new CallableHandler<Functor, Callable>(c));
   }
 
   template <typename Method>
   static Functor method(Method m) {
     return Functor(new MethodHandler<Functor, Method>(m));
@@ -644,9 +649,27 @@ template <typename R, typename P0, typen
 Functor<R, TLIST9(P0, P1, P2, P3, P4, P5, P6, P7, P8)>
 functor(R (*f)(P0, P1, P2, P3, P4, P5, P6, P7, P8)) {
   return Functor<R, TLIST9(P0, P1, P2, P3, P4, P5, P6, P7, P8)>::callable(f);
 }
 
 // [[[end]]]
 
 
+template <typename Result, class Params>
+nsRefPtr<nsRunnableFunctor<Result, Functor<Result, Params> > >
+Functor<Result, Params>::proxy(nsIThread* target, PRUint32 flags)
+{
+  static nsIThread* mainThread = nsnull;
+  if (!target && (mainThread || NS_SUCCEEDED(NS_GetMainThread(&mainThread))))
+      target = mainThread;
+
+  typedef nsRunnableFunctor<Result, Functor> Event;
+  nsRefPtr<Event> event(new Event(*this));
+
+  PRBool dispatched = target && NS_SUCCEEDED(target->Dispatch(event, flags));
+  NS_ASSERTION(dispatched, "Failed to dispatch event to thread.");
+
+  return event; // event->Wait() will block forever if Dispatch failed
+}
+
+
 #endif
diff --git a/xpcom/glue/Makefile.in b/xpcom/glue/Makefile.in
--- a/xpcom/glue/Makefile.in
+++ b/xpcom/glue/Makefile.in
@@ -120,16 +120,17 @@ SDK_HEADERS = \
 		nsDeque.h \
 		$(NULL)
 
 EXPORTS = \
 		nsThreadUtils.h \
 		nsProxyRelease.h \
 		nsXPTCUtils.h \
 		TList.h \
+		nsRunnableFunctor.h \
 		Functor.h \
 		Bind.h \
 		$(NULL)
 
 SDK_LIBRARY     =                        \
 		$(LIB_PREFIX)xpcomglue_s.$(LIB_SUFFIX) \
 		$(NULL)
 
diff --git a/xpcom/glue/nsRunnableFunctor.h b/xpcom/glue/nsRunnableFunctor.h
new file mode 100644
--- /dev/null
+++ b/xpcom/glue/nsRunnableFunctor.h
@@ -0,0 +1,99 @@
+#ifndef nsRunnableFunctor_h__
+#define nsRunnableFunctor_h__
+
+#include "nsISupports.h"
+#include "nsAutoPtr.h"
+#include "nsAutoLock.h"
+#include "nsIEventTarget.h"
+#include "nsThreadUtils.h"
+#include "nsProxyRelease.h"
+
+class ThreadSafeRunnable
+  : public nsIRunnable
+{
+  nsAutoRefCnt mRefCnt;
+public:
+  NS_IMPL_THREADSAFE_ISUPPORTS1(ThreadSafeRunnable, nsIRunnable)
+protected:
+  virtual ~ThreadSafeRunnable() {}
+};
+
+class nsARunnableFunctor
+  : public ThreadSafeRunnable
+{
+  PRMonitor* const mMon;
+  PRBool mRun;
+public:
+  nsARunnableFunctor()
+    : mMon(nsAutoMonitor::NewMonitor("nsARunnableFunctor"))
+    , mRun(PR_FALSE)
+  {}
+  void Wait()
+  {
+    nsAutoMonitor mon(mMon);
+    if (!mRun)
+      mon.Wait();
+  }
+  nsrefcnt Release()
+  {
+    if (NS_IsMainThread())
+      return ThreadSafeRunnable::Release();
+    nsIThread* mainThread = nsnull;
+    NS_GetMainThread(&mainThread);
+    if (mainThread)
+      NS_ProxyRelease(mainThread, this);
+    return 1;
+  }
+protected:
+  ~nsARunnableFunctor()
+  {
+    nsAutoMonitor::DestroyMonitor(mMon);
+  }
+  nsresult AfterRun()
+  {
+    nsAutoMonitor mon(mMon);
+    mRun = PR_TRUE;
+    mon.NotifyAll();
+    return NS_OK;
+  }
+};
+
+template <typename Result, typename Functor>
+class nsRunnableFunctor
+  : public nsARunnableFunctor
+{
+  Result mResult;
+  Functor mFunctor;
+public:
+  nsRunnableFunctor(Functor functor)
+    : mFunctor(functor)
+  {}
+  Result& Wait()
+  {
+    nsARunnableFunctor::Wait();
+    return mResult;
+  }
+  nsresult Run()
+  {
+    mResult = mFunctor();
+    return AfterRun();
+  }
+};
+
+template <typename Functor>
+class nsRunnableFunctor<void, Functor>
+  : public nsARunnableFunctor
+{
+  Functor mFunctor;
+public:
+  nsRunnableFunctor(Functor functor)
+    : mFunctor(functor)
+  {}
+  nsresult Run()
+  {
+    mFunctor();
+    return AfterRun();
+  }
+};
+
+#endif
