diff --git a/parser/html/nsHtml5MetaScannerCppSupplement.h b/parser/html/nsHtml5MetaScannerCppSupplement.h
--- a/parser/html/nsHtml5MetaScannerCppSupplement.h
+++ b/parser/html/nsHtml5MetaScannerCppSupplement.h
@@ -86,17 +86,17 @@ nsHtml5MetaScanner::tryCharset(nsString*
   CopyUTF16toUTF8(*charset, encoding);
   // XXX spec says only UTF-16
   if (encoding.LowerCaseEqualsASCII("utf-16") ||
       encoding.LowerCaseEqualsASCII("utf-16be") ||
       encoding.LowerCaseEqualsASCII("utf-16le") ||
       encoding.LowerCaseEqualsASCII("utf-32") ||
       encoding.LowerCaseEqualsASCII("utf-32be") ||
       encoding.LowerCaseEqualsASCII("utf-32le")) {
-    mCharset.Assign("utf-8");
+    mCharset.Assign("UTF-8");
     res = convManager->GetUnicodeDecoderRaw(mCharset.get(), getter_AddRefs(mUnicodeDecoder));
     if (NS_FAILED(res)) {
       NS_ERROR("Could not get decoder for UTF-8.");
       return PR_FALSE;
     }
     return PR_TRUE;
   }
   nsCAutoString preferred;
